### ISSUE TITLE
[New Concept Docs]: Dataclasses  in Python

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3109]
This issue describes how to implement the `dataclasses` in Python  **concept docs**.
You can find the related **concept exercise** issue [here](https://github.com/exercism/python/issues/3110)

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🎯   Goal

These concept docs are meant to teach an understanding/creation/use of `dataclasses` in Python. 

<br>

## 💡   Learning objectives

- more fully understand the uses (_and possible abuses_) of `classes` in Python
- learn some good use-cases for `dataclasses` over standard `classes`
- understand when and when **not to** use `dataclasses`
- create/use `dataclasses` and the `@dataclass` decorator
- learn some additional methods from the `dataclass` module
  - `dataclasses.fields`
  - `dataclasses.asdict` & `dataclasses.astuple`
  - `dataclasses.replace`
  - default factory functions
  - inheritance
- Contrast `dataclasses` with an alternative, `collections.namedtuple`
  - pros and cons of each method for creating data-focused `classes`
  - speed, memory, and typing use considerations

<br>

## 🚫  Out of scope

<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to open</em>)</summary>
<br>

- `class-inheritance`, `multiple-inheritance`, `__super()__`, class `mixins` (_these all have their own exercises_)
- `class-composition`  -- (_beyond the composition needed for the decorators in this exercise_)
- `collections` module, outside of what is needed to contrast `dataclasses` with `collections.namedtuple` (_the  `collections` module will get its own exercise_)
- `comprehensions`
- `coroutines`
- `decorators` outside of `@dataclass`
- `descriptors` (_these will get their own exercise_)
- `generators`
- `higher-order functions`
- using a `class` as a decorator, beyond the methods made available for `dataclasses`
- `type aliases`
- `type annotations`, beyond what is used in `dataclasses`

</details>

<br>

## 🤔  Concepts

<details>
<summary><b>Concepts & Related Concepts Covered</b> (<em>click to open</em>)</summary>
<br>

- `classes`
- `class attributes`
- `class members`
- `dataclasses`, `@dataclass`
- `decorators`
-  `collections.namedtuples()`

</details>

<br>

## ↩️   Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Prereqs  (<em>click to open</em>)</summary>
<br>

- `basics`
- `booleans`
- `classes`
- `class-customization`
-  `class-inheritance`
-  `class-composition`
- `comparisons`
- `decorators`
- `dicts`
-  `higher-order functions`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

</details>

<br>  

## 📚  Resources to refer to

<details>
<summary><b>Resources</b>  (<em>click to open</em>)</summary>
<br>

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [Python Docs: dataclasses](https://docs.python.org/3/library/dataclasses.html#module-dataclasses)
- [Python Docs: collections.namedtuple()](https://docs.python.org/3/library/collections.html#collections.namedtuple)
- [PEP 0557](https://www.python.org/dev/peps/pep-0557/)
- [PEP 0526](https://www.python.org/dev/peps/pep-0526/)
- [Real Python: Data Classes in Python 3.7+ (Guide)](https://realpython.com/python-data-classes/)
- [Dan Bader: Writing Clean Python with Namedtuples](https://dbader.org/blog/writing-clean-python-with-namedtuples)
- [Trey Hunner: Easier Classes - Python Classes without All the Cruft](https://treyhunner.com/easier-classes/nbpy2018.html)
- [SO:  What are data classes and how are they different?](https://stackoverflow.com/questions/47955263/what-are-data-classes-and-how-are-they-different-from-common-classes)
- [DataClass vs NamedTuple vs Object: A Battle of Performance in Python](https://medium.com/@jacktator/dataclass-vs-namedtuple-vs-object-for-performance-optimization-in-python-691e234253b9) -- _warning, this is a subscription service, so not great for `links.json`_
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)

</details>

<br>

##  📁  Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b> (<em>click to collapse</em>)</summary>
<table>
<td>
<br>

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>
</td>
</table>
</details>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.
